### ISSUE TITLE
Installing openh264 fails in containers and chroot

### ISSUE THREAD

**Issue by LaurentBonnaud** [flatpak/flatpak#3238]
## Linux distribution and version
Debian
## Flatpak version
1.4.3
## Description of the problem

Trying to install org.freedesktop.Platform.openh264 fails:
```
# flatpak update
Looking for updates?
        ID                                        Arch          Branch        Remote         Download
 1. [?] org.freedesktop.Platform.openh264         x86_64        19.08         flathub        593.6?kB / 593.4?kB
        ID                                        Arch          Branch        Remote         Download
 1. [?] org.freedesktop.Platform.openh264         x86_64        19.08         flathub        593.6?kB / 593.4?kB

Warning: While trying to apply extra data: apply_extra script failed, exit status 256
```

## Steps to reproduce
```
# docker run -it debian:sid
# apt install flatpak ca-certificates --no-install-recommends
# flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
# flatpak install org.blender.Blender
[...]
Warning: While trying to apply extra data: apply_extra script failed, exit status 256
```

**Comment by LaurentBonnaud** [comment]
The problem comes from the execution of this script:
`
/var/lib/flatpak/runtime/org.freedesktop.Platform.openh264/x86_64/19.08/563e6c1a71738559142a59f683bc43aebcc3e6f048413ebd9b64b4da5eb4597e/files/bin/apply_extra`

```
#!/usr/bin/bash

bzip2 -d libopenh264.so.bz2
mv libopenh264.so libopenh264.so.2.0.0
ldconfig $(pwd)
```

**Comment by LaurentBonnaud** [comment]
The error seems to come from bubblewrap when it tries to mount the root directory because when I run flatpak with "strace -f" I see the following:
```
[pid 27872] mount(NULL, "/", NULL, MS_REC|MS_SLAVE, NULL) = -1 EINVAL (Invalid argument)
[pid 27872] write(2, "bwrap: ", 7bwrap: )      = 7
[pid 27872] write(2, "Failed to make / slave", 22Failed to make / slave) = 22
[pid 27872] write(2, ": Invalid argument\n", 19: Invalid argument
```

**Comment by mwleeds** [comment]
It's quite possible that docker is not giving contained processes the privileges needed to create containers themselves. That's what bubblewrap does; see `drop_privs()` in `bubblewrap.c`.

**Comment by LaurentBonnaud** [comment]
The same problem exists with systemd containers and this command:
`systemd-nspawn -D DIRECTORY`

**Comment by mwleeds** [comment]
There's some relevant discussion on the mailing list: https://lists.freedesktop.org/archives/flatpak/2019-December/001840.html

**Comment by nanonyme** [comment]
Might work with privileged container.

**Comment by adamboutcher** [comment]
Is there anyway around this, I'm trying this within a chroot env.

**Comment by smcv** [comment]
> Is there anyway around this, I'm trying this within a chroot env.

The requirements that the Linux kernel imposes on us are quite complicated, and Flatpak is not in a position to change them: we can only work within what the kernel allows. Unfortunately, if one of those requirements is not met, the only information we are given is "Invalid argument". We are not told which argument was invalid or why.

https://github.com/containers/bubblewrap/issues/135 has some more technical details.

**Comment by LaurentBonnaud** [comment]
Thanks @smvc for pointing out this issue! I saw quite a lot of people trying to use nested containers and having a difficult time to do it (but some combinations do seem to work).

The root of the problem is that flatpak uses bubblewrap to execute the "apply_extra" script. Is this for security reasons (in case the script is malicious)?

Would it be possible for flatpak to execute the "apply_extra" script without using bubblewrap?  This behavior could be activated by a command line switch, for people (like me) who need to run flatpak in a container and accept the potential security risk.

**Comment by smcv** [comment]
> I saw quite a lot of people trying to use nested containers and having a difficult time to do it (but some combinations do seem to work)

Yes, nested containers usually don't work, but can be done if you are careful enough.

> The root of the problem is that flatpak uses bubblewrap to execute the "apply_extra" script. Is this for security reasons (in case the script is malicious)?

I think it's to prevent the `apply_extra` script from writing outside the area that it's meant to be allowed to modify, partly for security (in case the script or the extra data that it downloads is malicious) and partly to prevent mistakes. But I could be wrong.

> Would it be possible for flatpak to execute the "apply_extra" script without using bubblewrap?

I don't know, maybe? I mainly deal with the sandboxing part of Flatpak, and I don't know the package-management side of it so well. There are two questions, really:

1. Is it technically possible?
2. Is it an acceptable thing to allow from a correctness/safety point of view? What components or apps, outside Flatpak itself, are you trusting to be not malicious and not insecure if you do this?

The best way to find out whether it's technically possible would be for someone who wants the feature (maybe you?) to try to implement it and see what happens. Doing that would also give that someone a reasonable level of insight into whether it's something that makes sense to allow in terms of correctness/safety, and how best to document it to stop people shooting themselves in the foot, which they could discuss with the Flatpak maintainers on a pull request.

Flatpak has the concept of "triggers" which are run when an app is installed, to do things like updating the icon cache. There is already an option for whether to sandbox those, but it's currently a build-time option (Autotools `--disable-sandboxed-triggers` or Meson `-Dsandboxed_triggers=false`) rather than something you can pass to `flatpak install` or the libflatpak C APIs at runtime. Perhaps it would be valid to have a flag for whether to sandbox the triggers, and also whether to sandbox the `apply_extra` scripts?

When installing Flatpak apps system-wide, some of the work is done by the user who is running libflatpak code (via `flatpak install` or GNOME Software or similar) and some of the work is done by the `flatpak-system-helper` service, so it would be necessary to be careful about who trusts who to do what.

If your use-case is preparing an OS image that has some preinstalled Flatpak apps in it, then you might be able to make some simplifications, like for example only allowing uid 0 to do dangerous things.

**Comment by smcv** [comment]
> can pass to flatpak install or the libflatpak C APIs

(see also #5186)

**Comment by LaurentBonnaud** [comment]
> Is it technically possible?

It seems so. I looked at the `apply_extra_data()` function, and it uses `g_spawn_sync()` to execute `bwrap`.
It should be possible to execute the `files/bin/apply_extra` script directly instead.

**Comment by LaurentBonnaud** [comment]
> The best way to find out whether it's technically possible would be for someone who wants the feature (maybe you?) 

I am not the best person to do it:
 - I am already quite busy with my real job
 - I am not familiar with glib, which is heavily used in flatpak

**Comment by LaurentBonnaud** [comment]
> If your use-case is preparing an OS image that has some preinstalled Flatpak apps in it [...]

That is exactly my use case :>.

**Comment by noelmiller** [comment]
Using docker, there may be a workaround! (Results may vary, I haven't tested this beyond using it to install firefox and VLC.) 

If you run docker or podman using `--privileged` and you don't mount the FLATPAK_USER_DIR or FLATPAK_TRIGGERSDIR as volumes, installation will successfully complete with a single warning.

**Docker Command**
```
docker run --privileged --volume ./flatpaks:/flatpak-deps/output --rm flatpak-deps:latest org.mozilla.firefox
```

**Containerfile**
```
FROM ghcr.io/ublue-os/base-main:latest as flatpak-deps

RUN mkdir -p /flatpak-deps/output

COPY flatpak-deps.sh /flatpak-deps/flatpak-deps.sh

WORKDIR /flatpak-deps

VOLUME /flatpak-deps/output

ENTRYPOINT ["/bin/bash", "/flatpak-deps/flatpak-deps.sh"]
```

**Entry Point Script**
```
#!/bin/bash

set -exu

export FLATPAK_USER_DIR=$(mktemp -d)
export FLATPAK_TRIGGERSDIR=$(mktemp -d)
mkdir /var/tmp || true
chmod 1777 /var/tmp
flatpak config --user --set languages "*"
flatpak remote-add --user flathub https://flathub.org/repo/flathub.flatpakrepo
flatpak install --user -y "$@"
ostree refs --repo=${FLATPAK_USER_DIR}/repo
ostree refs --repo=${FLATPAK_USER_DIR}/repo | grep -E '^(flathub:)' | awk -F'flathub:' '{print $2}' | grep -E '^(app/|runtime/)' >/flatpak-deps/output/flatpaks
```

Feel free to give it a try! I'm using this as a step in a build process we are doing, just needed to get the dependencies to use for later.

**Comment by LaurentBonnaud** [comment]
I confirm that my original example works when I use `docker run --privileged`.
Thanks for this finding!
Unfortunately, I was using docker only as an example and my real use is with systemd-nspawn.
And I did not find a similar option in systemd-nspawn.
